design/one-pager-resource-usages.md

```diff
@@ -0,0 +1,346 @@
+# Modelling Resource Usage in Crossplane
+
+* Owner: Nic Cope (@negz)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Terminology
+
+* _External resource_. An actual resource that exists outside Kubernetes,
+  typically in the cloud. AWS RDS and GCP Cloud Memorystore instaces are
+  external resources.
+* _Managed resource_. The Crossplane representation of an external resource.
+  The `RDSInstance` and `CloudMemorystoreInstance` Kubernetes kinds are managed
+  resources. A managed resource models the satisfaction of a need; i.e. the need
+  for a Redis Cluster is satisfied by the allocation (aka binding) of a
+  `CloudMemoryStoreInstance`.
+* _Resource claim_. The Crossplane representation of a request for the
+  allocation of a managed resource. Resource claims typically represent the need
+  for a managed resource that implements a particular protocol. `MySQLInstance`
+  and `RedisCluster` are examples of resource claims.
+* _Resource class_. The Crossplane representation of the desired configuration
+  of a managed resource. Resource claims reference a resource class in order to
+  specify how they should be satisfied by a managed resource.
+* _Connection secret_. A Kubernetes `Secret` encoding all data required to
+  connect to (or consume) an external resource.
+* _Claimant_ or _consumer_. The Kubernetes representation of a process wishing
+  to connect to a managed resource, typically a `Pod` or some abstraction
+  thereupon such as a `Deployment` or `KubernetesApplication`.
+
+## Background
+
+Crossplane allows developers to manage and consume external resources via the
+Kubernetes API. Typically a few things must be known about an external resource
+in order to consume it: its URL, IP address, ports, credentials, etc. Managed
+resources expose the non-sensitive subset of this connection data via their
+`.status` object. All connection data, both sensitive and non-sensitive, is also
+exposed via a connection secret.
+
+Connection secret data has no schema; each managed resource controller specifies
+the keys and values that will be encoded in the `.data` object of its `Secret`.
+Constants such as `endpoint` and `password` serve to ensure there is some
+consistency in secret data keys across managed resources. Connection secrets are
+created at managed resource creation time in the same namespace as their
+managed resource (typically `crossplane-system`). Managed resources, which
+represent infrastructure external to Kubernetes, would not be namespaced at all
+if it were not for their need to persist connection data to a namespaced
+`Secret`. The `crossplane-system` namespace was originally envisioned to be the
+closest possible facsimile of cluster scoped resources given this dependency on
+namespaced `Secrets` per Crossplane issue [#92].
+
+Crossplane intends for consumers to declare their intent to consume a managed
+resource via a resource claim. Because resource claims typically exist in the
```

Comment:
Does this imply that using `Managed resource` is not a "recommended" way to provision and consume the `External resources`?

Reply:
I believe it is just denoting that while a `Managed resource` can be created directly, the power of Crossplane is that separation of concern created by the `Resource claim` + `Resource class` = `Managed resource` workflow

Reply:
> Does this imply that using `Managed resource` is not a "recommended" way to provision and consume the `External resources`?

@ichekrygin It implies that directly instantiating a managed resource, without ever creating a claim, is effectively unsupported. There would be no way to practically consume a managed resource without a claim, just like there is no way to practically consume a `PersistentVolume` without first creating a ` PersistentVolumeClaim` to signal your consumption.

Reply:
> just like there is no way to practically consume a PersistentVolume without first creating a PersistentVolumeClaim to signal your consumption.

Unless I misunderstand the statement above, but I think there is a way to consume Persistent Volumes w/out claims: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#volume-v1-core

Reply:
I think I see what is the issue w/ my understanding.  What I mean, is you can create volumes in pods w/out creating PVC/PV combo, by using directly supported persistent storage volumes like `awsElasticBlockStore` or `gcePersistentDisk`, etc.

Reply:
Putting aside comparisons to PV/PVC, do you think there's a use case for using managed resources directly, without a claim? Would there still be a case if we had 'strongly typed' claims (e.g. `RDSInstanceClaim`)?

Reply:
I think, yes, there will be. The basic use case I can think of, I want to provision and consume a managed resource w/out going into the lengths of `ResourceClass` and `ResourceClaim` - strongly typed or otherwise. 

Taxonomy of the direct resource provision/consume:
- CloudProvider <- CloudProviderSecret
- ManagedResource -> ManagedResourceConnectionSecret 
- Namespaces: 1 (all artifacts are live in "user" namespace)

Taxonomy of the Class/Claim resource provision/consume:
- CloudProvider <- CloudProviderSecret
- ResourceClass
- ResourceClaim -> ResourceClaimConnectionSecret
- ManagedResource -> ManagedResourceConnectionSecret
- Namespaces: 2
   - provider, resource class, managed resource, and their respective artifacts are deployed into "admin" namespace
   - claim and claim connection secret live in "user" namespace

It worth to note that adding  strongly type does not improve the taxonomy of the Class/Claim  option, however, it can make it more complicated if "Abstract" (Class/Claim) is built as an abstraction layer on top of the "Concrete" (Class/Claim), which would result in (hypothetical):
- CloudProvider <- CloudProviderSecret
- AbstractResourceClass
- AbstractResourceClaim -> AbstractResourceClaimConnectionSecret
- ConcreteResourceClass
- ConcreteResourceClaim -> ConcreteClaimConnectionSecret
- ManagedResource -> ManagedResourceConnectionSecret
- Namespaces: 2
   - provider, resource class, managed resource, and their respective artifacts are deployed into "admin" namespace
   - claim and claim connection secret live in "user" namespace

The main point I am trying to make is to entertain the idea of the "minimalistic" provision/consume scenario, i.e. the use case w/ least moving parts possible. 

If we think that in order to user crossplane, the user *must* buy into the separation of concerns, then I agree, the baseline use case is "Class/Claim->ManagedResource".

Reply:
> If we think that in order to user crossplane, the user must buy into the separation of concerns, then I agree, the baseline use case is "Class/Claim->ManagedResource".

I believe the claim has more value than simply separation of concerns - as with the PV/PVC pattern it's a concrete record that something is being "claimed" and thus should not be used by others. To my knowledge if we support resources being used directly (without a claim), and also support binding a claim to an existing resource, there's currently no way to distinguish between a managed resource that is available for binding, and a managed resource that is being used without a claim.

I feel requiring a resource claim simplifies the mental model for consumers:

* Bound to a claim - managed resource is in use.
* Not bound to a claim - managed resource is not in use.

Given the above, to me the marginal extra effort of creating a claim (and possibly class if a suitable one does not already exist) in order to create a managed resource is warranted.